File: leanreel/services/audit.py
"""压缩审计服务 — 构建、写入、读取 sidecar 文件"""
from __future__ import annotations

import datetime
import json
import platform
from pathlib import Path
from typing import Optional

from leanreel.domain.models import CompressionAudit


_LEANREEL_VERSION = "0.1.0"


def build_audit(
    task,
    ffmpeg_command: list[str],
    adaptive_cq_original: int = 0,
    adaptive_cq_adjusted: int = 0,
    adaptive_cq_reason: str = "",
) -> CompressionAudit:
    """从编码任务和策略构建完整审计快照。"""
    snap = task.snapshot
    strategy = task.strategy

    original = task.original_size if task.original_size > 0 else 1
    savings_bytes = task.original_size - task.compressed_size if task.compressed_size > 0 else 0
    savings_pct = round(savings_bytes / original * 100, 1) if original > 0 else 0.0

    video = getattr(strategy, "video", None)
    audio = getattr(strategy, "audio", None)
    sub = getattr(strategy, "subtitle", None)

    # 源文件音轨信息
    source_audio = []
    for i, a in enumerate(getattr(snap, "audio_tracks", []) or []):
        source_audio.append({
            "index": i,
            "codec": getattr(a, "codec", ""),
            "channels": getattr(a, "channels", 0),
            "language": getattr(a, "language", ""),
            "title": getattr(a, "title", ""),
            "is_commentary": getattr(a, "is_commentary", False),
        })

    # 源文件字幕信息
    source_subtitle = []
    for i, s in enumerate(getattr(snap, "subtitle_tracks", []) or []):
        source_subtitle.append({
            "index": i,
            "codec": getattr(s, "codec", ""),
            "language": getattr(s, "language", ""),
            "title": getattr(s, "title", ""),
            "is_forced": getattr(s, "is_forced", False),
        })

    # HDR 类型
    hdr_type = getattr(snap, "hdr_type", None)
    if hdr_type is not None and hasattr(hdr_type, "value"):
        source_hdr = hdr_type.value
    else:
        source_hdr = "SDR"

    # 计算实际持续时间
    started = getattr(task, "started_at", 0) or 0
    completed = getattr(task, "completed_at", 0) or 0
    if completed and started:
        task_duration = round(completed - started, 1)
    else:
        task_duration = 0.0

    # 状态
    task_status = getattr(task, "status", None)
    if task_status is not None and hasattr(task_status, "value"):
        status_str = task_status.value
    else:
        status_str = str(task_status) if task_status else "pending"

    return CompressionAudit(
        library_folder_id=getattr(snap, "library_folder_id", 0),
        relative_path=getattr(snap, "relative_path", ""),
        source_path=task.input_path,
        source_size_bytes=getattr(snap, "size_bytes", 0),
        source_mtime=getattr(snap, "file_mtime", 0.0),
        source_codec=getattr(snap, "video_codec", ""),
        source_width=getattr(snap, "video_width", 0),
        source_height=getattr(snap, "video_height", 0),
        source_pix_fmt=getattr(snap, "pix_fmt", ""),
        source_bitrate_bps=getattr(snap, "bitrate_bps", 0),
        source_duration_seconds=getattr(snap, "duration_seconds", 0.0),
        source_frame_rate=getattr(snap, "frame_rate", ""),
        source_hdr=source_hdr,
        source_audio=source_audio,
        source_subtitle=source_subtitle,

        output_path=task.output_path,
        output_size_bytes=task.compressed_size,
        savings_bytes=savings_bytes,
        savings_pct=savings_pct,

        strategy_name=getattr(strategy, "name", task.strategy_name),
        encoder=getattr(video, "encoder", "") if video else "",
        crf=getattr(video, "crf", 0) if video else 0,
        cq=getattr(video, "cq", 0) if video else 0,
        preset=getattr(video, "preset", "") if video else "",
        pix_fmt=getattr(video, "pix_fmt", "") if video else "",
        audio_mode=getattr(audio, "mode", "") if audio else "",
        sub_mode=getattr(sub, "mode", "") if sub else "",

        ffmpeg_command=list(ffmpeg_command),
        adaptive_cq_original=adaptive_cq_original,
        adaptive_cq_adjusted=adaptive_cq_adjusted,
        adaptive_cq_reason=adaptive_cq_reason,
        started_at=_iso(task.started_at) if started else "",
        completed_at=_iso(task.completed_at) if completed else "",
        duration_seconds=task_duration,
        status=status_str,

        ffmpeg_version=_ffmpeg_version(),
        dovi_tool_version=_dovi_version(),
        leanreel_version=_LEANREEL_VERSION,
        platform=platform.platform(),
    )


def _iso(timestamp: float) -> str:
    if not timestamp:
        return ""
    return datetime.datetime.fromtimestamp(timestamp).isoformat()


def _ffmpeg_version() -> str:
    from leanreel.executor.ffmpeg_builder import get_ffmpeg_version
    return get_ffmpeg_version()


def _dovi_version() -> str:
    from leanreel.executor.dovi import get_dovi_tool_version
    return get_dovi_tool_version()


def write_sidecar(audit: CompressionAudit) -> str:
    """将审计快照写入 sidecar JSON 文件。返回写入的路径。失败时返回空字符串。"""
    sidecar_path = _sidecar_path(audit.output_path)
    data = _audit_to_dict(audit)
    try:
        sidecar_path.parent.mkdir(parents=True, exist_ok=True)
        sidecar_path.write_text(json.dumps(data, ensure_ascii=False, indent=2), encoding="utf-8")
        return str(sidecar_path)
    except Exception:
        import traceback
        traceback.print_exc()
        return ""


def read_sidecar(filepath: str) -> Optional[CompressionAudit]:
    """从 sidecar JSON 文件读取审计快照。失败返回 None。"""
    try:
        data = json.loads(Path(filepath).read_text(encoding="utf-8"))
        return _dict_to_audit(data)
    except Exception:
        return None


def find_sidecars_for_source(source_path: str) -> list[str]:
    """在源文件同目录下查找对应的 sidecar 文件。"""
    source = Path(source_path)
    pattern = f"{source.stem}_zcompressed.*.leanreel.json"
    return sorted(
        str(p) for p in source.parent.glob(pattern)
        if p.is_file()
    )


def _sidecar_path(output_path: str) -> Path:
    return Path(output_path + ".leanreel.json")


def _audit_to_dict(audit: CompressionAudit) -> dict:
    result = {
        "version": "1.0",
        "tool": f"LeanReel {audit.leanreel_version}",
        "source": {
            "path": audit.source_path,
            "file_name": Path(audit.source_path).name if audit.source_path else "",
            "size_bytes": audit.source_size_bytes,
            "mtime": audit.source_mtime,
            "video": {
                "codec": audit.source_codec,
                "width": audit.source_width,
                "height": audit.source_height,
                "pix_fmt": audit.source_pix_fmt,
                "bitrate_bps": audit.source_bitrate_bps,
                "duration_seconds": audit.source_duration_seconds,
                "frame_rate": audit.source_frame_rate,
                "hdr_type": audit.source_hdr,
                "color_primaries": audit.source_color_primaries,
                "color_transfer": audit.source_color_transfer,
                "color_space": audit.source_color_space,
            },
            "audio": audit.source_audio,
            "subtitle": audit.source_subtitle,
        },
        "output": {
            "path": audit.output_path,
            "size_bytes": audit.output_size_bytes,
            "savings_bytes": audit.savings_bytes,
            "savings_pct": audit.savings_pct,
        },
        "strategy": {
            "name": audit.strategy_name,
            "video": {
                "encoder": audit.encoder,
                "crf": audit.crf or audit.cq,
                "preset": audit.preset,
                "pix_fmt": audit.pix_fmt,
            },
            "audio": {"mode": audit.audio_mode},
            "subtitle": {"mode": audit.sub_mode},
        },
        "execution": {
            "command": audit.ffmpeg_command,
            "adaptive_cq": {
                "original": audit.adaptive_cq_original,
                "adjusted": audit.adaptive_cq_adjusted,
                "reason": audit.adaptive_cq_reason,
            } if audit.adaptive_cq_original else {},
            "started_at": audit.started_at,
            "completed_at": audit.completed_at,
            "duration_seconds": audit.duration_seconds,
            "status": audit.status,
        },
        "environment": {
            "ffmpeg_version": audit.ffmpeg_version,
            "dovi_tool_version": audit.dovi_tool_version,
            "leanreel_version": audit.leanreel_version,
            "platform": audit.platform,
        },
    }
    return result


def _dict_to_audit(data: dict) -> CompressionAudit:
    src = data.get("source", {})
    src_video = src.get("video", {})
    out = data.get("output", {})
    strat = data.get("strategy", {})
    strat_video = strat.get("video", {})
    exe = data.get("execution", {})
    adaptive_cq = exe.get("adaptive_cq", {})
    env = data.get("environment", {})

    # 策略视频参数中 crf/cq 二选一
    cq_val = strat_video.get("crf", strat_video.get("cq", 0))

    return CompressionAudit(
        source_path=src.get("path", ""),
        source_size_bytes=src.get("size_bytes", 0),
        source_mtime=src.get("mtime", 0.0),
        source_codec=src_video.get("codec", ""),
        source_width=src_video.get("width", 0),
        source_height=src_video.get("height", 0),
        source_pix_fmt=src_video.get("pix_fmt", ""),
        source_bitrate_bps=src_video.get("bitrate_bps", 0),
        source_duration_seconds=src_video.get("duration_seconds", 0.0),
        source_frame_rate=src_video.get("frame_rate", ""),
        source_hdr=src_video.get("hdr_type", "SDR"),
        source_audio=src.get("audio", []),
        source_subtitle=src.get("subtitle", []),
        output_path=out.get("path", ""),
        output_size_bytes=out.get("size_bytes", 0),
        savings_bytes=out.get("savings_bytes", 0),
        savings_pct=out.get("savings_pct", 0.0),
        strategy_name=strat.get("name", ""),
        encoder=strat_video.get("encoder", ""),
        cq=cq_val,
        preset=strat_video.get("preset", ""),
        pix_fmt=strat_video.get("pix_fmt", ""),
        audio_mode=strat.get("audio", {}).get("mode", ""),
        sub_mode=strat.get("subtitle", {}).get("mode", ""),
        ffmpeg_command=exe.get("command", []),
        adaptive_cq_original=adaptive_cq.get("original", 0) if adaptive_cq else 0,
        adaptive_cq_adjusted=adaptive_cq.get("adjusted", 0) if adaptive_cq else 0,
        adaptive_cq_reason=adaptive_cq.get("reason", "") if adaptive_cq else "",
        started_at=exe.get("started_at", ""),
        completed_at=exe.get("completed_at", ""),
        duration_seconds=exe.get("duration_seconds", 0.0),
        status=exe.get("status", "pending"),
        ffmpeg_version=env.get("ffmpeg_version", ""),
        dovi_tool_version=env.get("dovi_tool_version", ""),
        leanreel_version=env.get("leanreel_version", ""),
        platform=env.get("platform", ""),
    )

File: src-tauri/src/services/audit.rs
use super::time_utils::local_now;
use crate::domain::models::{
    CompressionAudit, DvProfile, FileSnapshot, HdrType, Strategy, VideoCodec,
};
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::Path;
use std::process::Command;

// ── Legacy AuditRecord (11-field, kept for backward compatibility) ────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuditRecord {
    pub timestamp: String,
    pub source_path: String,
    pub output_path: String,
    pub source_size_bytes: u64,
    pub output_size_bytes: u64,
    pub source_codec: String,
    pub output_codec: String,
    pub strategy_name: String,
    pub duration_seconds: f64,
    pub success: bool,
    pub error_message: String,
}

/// 写入审计侧挂文件（.leanreel.json）
pub fn write_audit_sidecar(sidecar_path: &Path, record: &AuditRecord) -> Result<(), String> {
    let json =
        serde_json::to_string_pretty(record).map_err(|e| format!("序列化审计记录失败: {}", e))?;
    fs::write(sidecar_path, json).map_err(|e| format!("写入审计文件失败: {}", e))?;
    Ok(())
}

/// 读取审计侧挂文件
pub fn read_audit_sidecar(sidecar_path: &Path) -> Result<AuditRecord, String> {
    let json = fs::read_to_string(sidecar_path).map_err(|e| format!("读取审计文件失败: {}", e))?;
    serde_json::from_str(&json).map_err(|e| format!("解析审计文件失败: {}", e))
}

// ── CompressionAudit (35-field) ──────────────────────────────────────────────

/// Build a complete audit record from encoding job parameters and result.
///
/// All fields (50+) are populated from real data — FileSnapshot provides source metadata,
/// Strategy provides encoding configuration, and the remaining parameters capture
/// runtime output details and execution status.
/// H-029: Added 10+ missing fields: source_pix_fmt, source_frame_rate,
/// source_color_primaries, source_color_transfer, source_color_space,
/// platform, adaptive_cq_original, adaptive_cq_adjusted, adaptive_cq_reason.
pub fn build_audit(
    snapshot: &FileSnapshot,
    output_path: &Path,
    output_size: u64,
    output_codec: &str,
    strategy: &Strategy,
    duration_ms: u64,
    success: bool,
    error: &str,
    ffmpeg_command: &str,
) -> CompressionAudit {
    let source_size = snapshot.size_bytes as u64;

    let savings_pct = if source_size > 0 {
        ((source_size as f64 - output_size as f64) / source_size as f64) * 100.0
    } else {
        0.0
    };

    // ── derive source metadata from snapshot ──────────────────────────────

    let source_codec = video_codec_to_str(&snapshot.video_codec);

    let source_hdr = hdr_type_to_str(&snapshot.hdr_type);

    let has_dolby_vision = matches!(snapshot.hdr_type, HdrType::DolbyVision { .. });

    // ── H-029: platform detection ─────────────────────────────────────────
    let platform = format!("{}-{}", std::env::consts::OS, std::env::consts::ARCH);

    // ── build the full audit record ───────────────────────────────────────

    CompressionAudit {
        // Version & tools
        leanreel_version: option_env!("CARGO_PKG_VERSION")
            .unwrap_or("unknown")
            .to_string(),
        ffmpeg_version: get_ffmpeg_version(),
        dovi_tool_version: get_dovi_tool_version(),

        // Source file info — all from FileSnapshot
        source_path: snapshot.relative_path.clone(),
        source_size_bytes: source_size,
        source_codec,
        source_width: snapshot.video_width,
        source_height: snapshot.video_height,
        source_hdr,
        source_duration_seconds: snapshot.duration_seconds,
        source_bitrate_bps: snapshot.bitrate_bps,
        source_audio_count: snapshot.audio_tracks.len(),
        source_subtitle_count: snapshot.subtitle_tracks.len(),
        // H-029: extended probe fields from snapshot
        source_pix_fmt: snapshot.pix_fmt.clone(),
        source_frame_rate: snapshot.frame_rate.clone(),
        source_color_primaries: snapshot.color_primaries.clone(),
        source_color_transfer: snapshot.color_transfer.clone(),
        source_color_space: snapshot.color_space.clone(),
        source_mtime: snapshot.file_mtime,

        // Output file info
        output_path: output_path.to_string_lossy().to_string(),
        output_size_bytes: output_size,
        output_codec: output_codec.to_string(),
        savings_pct,
        size_delta_bytes: snapshot.size_bytes - output_size as i64,

        // Strategy details — all from Strategy
        strategy_name: strategy.name.clone(),
        encoder: strategy.video.encoder.clone(),
        cq_value: strategy.video.cq,
        crf_value: strategy.video.crf,
        preset: strategy.video.preset.clone(),
        pix_fmt: strategy.video.pix_fmt.clone(),
        audio_mode: strategy.audio.mode.clone(),
        sub_mode: strategy.subtitle.mode.clone(),

        // Execution
        duration_ms,
        success,
        error_message: error.to_string(),
        ffmpeg_command: ffmpeg_command.to_string(),
        status: if success {
            "completed".into()
        } else {
            "failed".into()
        },
        stage: String::new(),
        progress: 0.0,
        // H-029: adaptive CQ fields (populated by caller, defaults to 0/empty)
        adaptive_cq_original: 0,
        adaptive_cq_adjusted: 0,
        adaptive_cq_reason: String::new(),

        // Environment
        timestamp: local_now(),
        completed_at: local_now(),
        source_deleted: false,
        has_dolby_vision,
        dv_handling: strategy.hdr.dv_handling.clone(),
        batch_id: String::new(),
        platform,
    }
}

/// Query ffmpeg for its version string (first line of `ffmpeg -version`).
fn get_ffmpeg_version() -> String {
    Command::new("ffmpeg")
        .arg("-version")
        .output()
        .ok()
        .and_then(|o| String::from_utf8(o.stdout).ok())
        .and_then(|s| s.lines().next().map(|l| l.to_string()))
        .unwrap_or_default()
}

/// Query dovi_tool for its version string (first line of `dovi_tool --version`).
fn get_dovi_tool_version() -> String {
    Command::new("dovi_tool")
        .arg("--version")
        .output()
        .ok()
        .and_then(|o| String::from_utf8(o.stdout).ok())
        .and_then(|s| s.lines().next().map(|l| l.to_string()))
        .unwrap_or_default()
}

/// Convert VideoCodec enum to a canonical short codec string.
fn video_codec_to_str(codec: &VideoCodec) -> String {
    match codec {
        VideoCodec::H264 => "h264".into(),
        VideoCodec::Hevc => "hevc".into(),
        VideoCodec::Av1 => "av1".into(),
        VideoCodec::Vp9 => "vp9".into(),
        VideoCodec::Mpeg2 => "mpeg2".into(),
        VideoCodec::Vc1 => "vc1".into(),
        VideoCodec::Unknown(s) => s.clone(),
    }
}

/// Convert HdrType enum to a human-readable string for audit records.
fn hdr_type_to_str(hdr: &HdrType) -> String {
    match hdr {
        HdrType::Sdr => "SDR".into(),
        HdrType::Hdr10 => "HDR10".into(),
        HdrType::Hdr10Plus => "HDR10+".into(),
        HdrType::DolbyVision { profile } => match profile {
            DvProfile::Profile5 => "DolbyVision:Profile5".into(),
            DvProfile::Profile7 => "DolbyVision:Profile7".into(),
            DvProfile::Profile8_1 => "DolbyVision:Profile8_1".into(),
            DvProfile::Profile8_4 => "DolbyVision:Profile8_4".into(),
        },
    }
}

/// Write audit sidecar JSON file alongside output
pub fn write_sidecar(output_path: &Path, audit: &CompressionAudit) -> Result<(), String> {
    let sidecar_path = std::path::PathBuf::from(format!("{}.leanreel.json", output_path.display()));
    let json = serde_json::to_string_pretty(audit).map_err(|e| format!("序列化审计失败: {}", e))?;
    fs::write(&sidecar_path, json).map_err(|e| format!("写入审计文件失败: {}", e))?;
    Ok(())
}
